// @pchmn/expo-material3-theme/build/utils/createMaterial3Theme.js
import { argbFromHex, themeFromSourceColor } from '@importantimport/material-color-utilities';
import color from 'color';
const opacity = {
    level1: 0.08,
    level2: 0.12,
    level3: 0.16,
    level4: 0.38,
};
const elevations = ['transparent', 0.05, 0.08, 0.11, 0.12, 0.14];
export function createThemeFromSystemSchemes(schemes) {
    const { light, dark, palettes } = generateSchemesFromSourceColor(schemes.light.primary);
    schemes = {
        light: { ...light, ...schemes.light },
        dark: { ...dark, ...schemes.dark },
    };
    return {
        light: { ...schemes.light, ...generateMissingFields(schemes.light, palettes) },
        dark: { ...schemes.dark, ...generateMissingFields(schemes.dark, palettes) },
    };
}
export function createThemeFromSourceColor(sourceColor) {
    const { light, dark, palettes } = generateSchemesFromSourceColor(sourceColor);
    return {
        light: { ...light, ...generateMissingFields(light, palettes) },
        dark: { ...dark, ...generateMissingFields(dark, palettes) },
    };
}
function generateMissingFields(scheme, palettes) {
    const elevation = elevations.reduce((acc, value, index) => ({
        ...acc,
        [`level${index}`]: index === 0 ? value : color(scheme.surface).mix(color(scheme.primary), Number(value)).hex(),
    }), {});
    const customColors = {
        surfaceDisabled: color(scheme.onSurface).alpha(opacity.level2).rgb().string(),
        onSurfaceDisabled: color(scheme.onSurface).alpha(opacity.level4).rgb().string(),
        backdrop: color(palettes.neutralVariant.tone(20)).alpha(0.4).rgb().string(),
    };
    return { elevation, ...customColors };
}
function generateSchemesFromSourceColor(sourceColor) {
    const { schemes, palettes } = themeFromSourceColor(argbFromHex(sourceColor));
    return {
        light: transformScheme(schemes.light),
        dark: transformScheme(schemes.dark),
        palettes,
    };
}
function transformScheme(scheme) {
    const jsonScheme = scheme.toJSON();
    return Object.entries(jsonScheme).reduce((acc, [key, value]) => {
        return {
            ...acc,
            [key]: color(value).hex(),
        };
    }, {});
}
